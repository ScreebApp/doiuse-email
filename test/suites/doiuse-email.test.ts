import { outdent } from 'outdent';
import { describe, expect, test } from 'vitest';

import { doIUseEmail } from '~/index.js';

describe('doIUseEmail() works', () => {
	test('works with blank email template', () => {
		const code = outdent`
			<!doctype html>
			<html>
				<body>
				</body>
			</html>
		`;
		const result = doIUseEmail(code, { emailClients: ['gmail.*'] });
		expect(result.success).toEqual(true);
		expect(result).toMatchSnapshot();
	});

	test('parses inline CSS', () => {
		const code = outdent`
			<!doctype html>
			<html>
				<body>
					<div style='background-color: orange'></div>
				</body>
			</html>
		`;
		const result = doIUseEmail(code, { emailClients: ['gmail.*'] });
		expect(result.success).toEqual(true);
		expect(result).toMatchSnapshot();
	});

	test('should fail on unsupported <style> features', () => {
		// `filter` isn't supported by gmail.desktop-webmail: https://www.caniemail.com/features/css-filter/
		const code = outdent`
			<!doctype html>
			<html>
				<head>
					<style>
						.container {
							filter: blur(50%);
						}
					</style>
				</head>
				<body>
					<div class='container' style='background-color: orange'></div>
				</body>
			</html>
		`;
		const result = doIUseEmail(code, {
			emailClients: ['gmail.desktop-webmail'],
		});
		expect(result.success).toEqual(false);
		expect(result).toMatchSnapshot();
	});

	test('should fail on unsupported inline-style features', () => {
		// `flex-direction: column` isn't supported by Gmail: https://www.caniemail.com/features/css-flex-direction/
		const code = outdent`
			<!doctype html>
			<html>
				<body>
					<div style='flex-direction: column'></div>
				</body>
			</html>
		`;
		const result = doIUseEmail(code, {
			emailClients: ['gmail.desktop-webmail'],
		});
		expect(result.success).toEqual(false);
		expect(result).toMatchSnapshot();
	});
});
