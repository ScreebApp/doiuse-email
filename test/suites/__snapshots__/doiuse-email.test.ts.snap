// Vitest Snapshot v1

exports[`doIUseEmail() works > parses inline CSS 1`] = `
{
  "notes": [
    "Note about \`<body> element\` support for \`gmail.desktop-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.ios\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.android\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.mobile-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
  ],
  "success": true,
}
`;

exports[`doIUseEmail() works > should fail on unsupported <style> features 1`] = `
{
  "errors": [
    "filter is not supported by gmail.desktop-webmail",
  ],
  "notes": [
    "Note about \`<style> element\` support for \`gmail.desktop-webmail\`: Partial. Not supported inside the \`<body>\`.",
    "Note about \`<body> element\` support for \`gmail.desktop-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
  ],
  "success": false,
}
`;

exports[`doIUseEmail() works > should fail on unsupported inline-style features 1`] = `
{
  "errors": [
    "flex-direction:column is not supported by gmail.desktop-webmail",
  ],
  "notes": [
    "Note about \`<body> element\` support for \`gmail.desktop-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
  ],
  "success": false,
}
`;

exports[`doIUseEmail() works > works with blank email template 1`] = `
{
  "notes": [
    "Note about \`<body> element\` support for \`gmail.desktop-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.ios\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.android\`: Partial. Replaced by a \`<div>\` with supported attributes.",
    "Note about \`<body> element\` support for \`gmail.mobile-webmail\`: Partial. Replaced by a \`<div>\` with supported attributes.",
  ],
  "success": true,
}
`;
